pkg/virt-controller/services/renderresources.go

```diff
@@ -0,0 +1,554 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/util/hardware"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+)
+
+type ResourceRendererOption func(renderer *ResourceRenderer)
+
+type ResourceRenderer struct {
+	vmLimits           k8sv1.ResourceList
+	vmRequests         k8sv1.ResourceList
+	calculatedLimits   k8sv1.ResourceList
+	calculatedRequests k8sv1.ResourceList
+}
+
+type resourcePredicate func(*v1.VirtualMachineInstance) bool
+
+type VMIPredicates struct {
+	resourceRules []VMIResourceRule
+	vmi           *v1.VirtualMachineInstance
+}
+
+type VMIResourceRule struct {
+	predicate resourcePredicate
+	option    ResourceRendererOption
+}
+
+func not(p resourcePredicate) resourcePredicate {
+	return func(instance *v1.VirtualMachineInstance) bool {
+		return !p(instance)
+	}
+}
+func NewVMIResourceRule(p resourcePredicate, option ResourceRendererOption) VMIResourceRule {
+	return VMIResourceRule{predicate: p, option: option}
+}
+
+func NewVMIPredicates(vmi *v1.VirtualMachineInstance, clusterConfig *virtconfig.ClusterConfig, networkResources map[string]string) VMIPredicates {
+	memoryOverhead := GetMemoryOverhead(vmi, clusterConfig.GetClusterCPUArch())
```

Comment:
Not sure if this is related to this PR, but the CPU architecture can be different between nodes

Reply:
I guess a follow-up PR should then address this here (implement per node architecture considerations plus drop the `clusterConfig.GetClusterCPUArch()` attribute).

Since I am only re-writing existing code, I agree - it is not related to this PR.